### PR TITLE
Add Makefile, configuration files and scripts

### PR DESCRIPTION
**Pull request by razvand**
Update the README with instructions on setting up, configuring, building and running the Nginx app with Unikraft. As part of it, add Makefile files, configuration files and scripts; they are referenced in the `README.md` file.

Current instructions cover the use of QEMU/KVM platform (both on x86_64 and on AArch64) and GCC. Firecraker/KVM, Xen, Linux platforms and Clang are not yet documented.

### CODE REVIEW

File: config-qemu-x86_x64
Comment (StefanJum) [inline config-qemu-x86_x64:1]: Rename this to `config-qemu-x86_64` (not `x86-x64`).

File: README.md [hunk 1]
```diff
@@ -1,93 +1,202 @@
...
+
+     ```console
+     $ cd ../
+     ```
+
+     Use the `tree` command to inspect the contents of the `.unikraft/` directory.
+     It should print something like this:
+
+     ```console
+     $ tree --charset=ascii -F -L 2 .unikraft/
```
Comment (StefanJum) [inline README.md]:
```suggestion
     $ tree -F -L 2 .unikraft/
```

File: README.md [hunk 2]
```diff
@@ -1,93 +1,202 @@
...
+
+  1. While inside the `.unikraft/` directory, create the `libs/` directory:
+
+     ```console
+     $ mkdir libs
+     ```
+
+  1. While inside the `.unikraft/` directory, clone the library repositories in the `libs/` directory:
+
+     ```console
```
Comment (StefanJum) [inline README.md:108]:
```suggestion
     ```console
     $ cd libs/
```

File: README.md [hunk 3]
```diff
@@ -1,93 +1,202 @@
...
+
+     $ git clone https://github.com/unikraft/lib-musl musl
+
+     $ git clone https://github.com/unikraft/lib-lwip lwip
+     ```
+
+  1. Get back to the application directory:
+
+     ```console
+     $ cd ../
```
Comment (StefanJum) [inline README.md:119]:
```suggestion
     $ cd ../../
```

File: README.md [hunk 4]
```diff
@@ -1,93 +1,202 @@
...
+Supported architectures are x86_64 and AArch64.
+
+Use the corresponding the configuration files (`config-...`), according to your choice of platform and architecture.
+
+### QEMU x86_64
+
+Use the `config-qemu-x86_64` configuration file together with `make defconfig` to create the configuration file:
+
+```console
+$ UK_DEFCONFIG=$(pwd)/config-qemu-x84_64 make defconfig
```
Comment (StefanJum) [inline README.md]:
```suggestion
$ UK_DEFCONFIG=$(pwd)/config-qemu-x86_64 make defconfig
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
@razvand first round

File: run-qemu-aarch64.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+sudo ip address add 172.44.0.1/24 dev virbr0
+sudo ip link set dev virbr0 up
+
+sudo /usr/bin/qemu-system-aarch64 \
+    -fsdev local,id=myid,path=$(pwd)/fs0,security_model=none \
+    -device virtio-9p-pci,fsdev=myid,mount_tag=fs0,disable-modern=on,disable-legacy=off \
+    -netdev bridge,id=en0,br=virbr0 -device virtio-net-pci,netdev=en0 \
+    -append "netdev.ipv4_addr=172.44.0.2 netdev.ipv4_gw_addr=172.44.0.1 netdev.ipv4_subnet_mask=255.255.255.0 --" \
+    -kernel build/nginx_qemu-arm64 -nographic \
+    -machine virt -cpu cortex-a57
```
Comment (StefanJum) [inline run-qemu-aarch64.sh]:
```suggestion
    -machine virt -cpu max
```

File: README.md [hunk 6]
```diff
@@ -1,93 +1,307 @@
...
 ```
-$ ip a s virbr0
-420: virbr0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
-    link/ether 3a:3e:88:e6:a1:e4 brd ff:ff:ff:ff:ff:ff
-    inet 172.44.0.1/24 scope global virbr0
-       valid_lft forever preferred_lft forever
-    inet6 fe80::383e:88ff:fee6:a1e4/64 scope link
-       valid_lft forever preferred_lft forever
+
+At the end of the build command, the `app-nginx_qemu-x86_64` unikernel image is generated.
```
Comment (StefanJum) [inline README.md]:
```suggestion
At the end of the build command, the `nginx_qemu-x86_64` unikernel image is generated.
```

File: README.md [hunk 7]
```diff
@@ -1,93 +1,307 @@
...
-[...]
-0: Set IPv4 address 172.44.0.2 mask 255.255.255.0 gw 172.44.0.1
-en0: Added
-en0: Interface is up
+$ make CROSS_COMPILE=~/packages/gcc-arm-11.2-2022.02-x86_64-aarch64-none-linux-gnu/bin/aarch64-none-linux-gnu- -j $(nproc)
 [...]
+  LD      app-nginx_qemu-arm64.dbg
+  UKBI    app-nginx_qemu-arm64.dbg.bootinfo
+  SCSTRIP app-nginx_qemu-arm64
+  GZ      app-nginx_qemu-arm64.gz
```
Comment (StefanJum) [inline README.md]:
```suggestion
  LD      nginx_qemu-arm64.dbg
  UKBI    nginx_qemu-arm64.dbg.bootinfo
  SCSTRIP nginx_qemu-arm64
  GZ      nginx_qemu-arm64.gz
```

File: README.md [hunk 8]
```diff
@@ -1,93 +1,307 @@
...
+make[1]: Leaving directory '/media/razvan/c4f6765a-efa5-4ebd-9cf0-7da9908a0189/razvan/unikraft/solo/app-nginx/.unikraft/unikraft
 ```
 
-The boot message confirms the assigning of the `172.44.0.2/24` IP address to the virtual machine.
-We use `wget` to validate it's working properly and we are able to get the `index.html` file:
+Be sure to pass a string ending in `-`;
+that is, DO NOT fill the command name: `aarch64-none-linux-gnu-gcc`;
+simply end it with `aarch64-none-linux-gnu-`.
+
+Similarly to x86_64, at the end of the build command, the `app-nginx_qemu-arm64` unikernel image is generated.
```
Comment (StefanJum) [inline README.md]:
```suggestion
Similarly to x86_64, at the end of the build command, the `nginx_qemu-arm64` unikernel image is generated.
```

File: fs0/nginx/logs/.gitignore
Comment (StefanJum) [inline fs0/nginx/logs/.gitignore:1]: What is with this file?

**Review by StefanJum — CHANGES_REQUESTED** [review]

File: fs0/nginx/logs/.gitignore
Comment (razvand) [reply fs0/nginx/logs/.gitignore:1]: Placeholder to still keep "alive" the `nginx/logs/` directory.

File: README.md [hunk 9]
```diff
@@ -1,93 +1,312 @@
...
+* `make clean`: cleans all actual build output files (binary files, including the unikernel image)
+* `make properclean`: removes the entire `build/` directory
+* `make distclean`: removes the entire `build/` directory **and** the `.config` file
+
+Typically, you would use `make properclean` to remove all build artifacts, but keep the configuration file.
+
+### QEMU x86_64
+
+Building for QEMU x86_64 assumes you did the QEMU x86_64 configuration step above.
+Build the Unikraft Nginx image for QEMU AArch64 by using the command below:
```
Comment (StefanJum) [inline README.md]:
```suggestion
Build the Unikraft Nginx image for QEMU x86_64 by using the command below:
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
Looks fine besides the comment

**Review by StefanJum — APPROVED** [review]
All good, thanks.

Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Review by John-Ted — APPROVED** [review]
All good!

Reviewed-by: Ioan-Teodor Teugea ioan_teodor.teugea@stud.acs.upb.ro

**Pull request closed without merging**